README.md
# prog-distrib

## Introduction

Ces sources sont une copie des sources pcsc-lite permettant la communication entre le système et les lecteurs de puces NFC, avec ajout d'une fonction et de ses fichiers d'entête spécifique à notre projet objet connecté.

Tous les ajouts spécifiques au projet du distributeur (donc tout ce qui n'est pas natif aux sources pcsc-lite) se trouvent dans le dossier `include`.
Ajout également d'une ligne de code dans le programme principal `pcsc_scan.c` nommée `onTagOn()` visible à la ligne 895.

> Ce README ne portera que sur cette fonctionnalité et tout ce qui en découle. Pour plus d'informations sur le fonctionnement des librairies pcsc, je vous invite à consulter la page git du paquet natif.

## Mise en place programme

Un script automatise la mise en place sur le Drive.


**TODO**
Ajouter la librairie wiringPi à la compilation et les programmes `run()` et  `stop()` pour faire fonctionner le moteur.


## Le script auto

- install gawk
- install mariadb-server
- install libmariadb-dev
- install libmariadb-dev-compat
- install pcscd
- install libpcsclite-dev
- install automake



Changelog
commit c6e9fa746d00e08fc71c57c79e305669ff2e3106 (HEAD -> master, zotac/master, origin/master)
Author: Ludovic Rousseau <ludovic.rousseau@free.fr>
Date:   Sat Jan 29 18:46:30 2022 +0100

    Release 1.6.0
    
    Signed-off-by: Ludovic Rousseau <ludovic.rousseau@free.fr>

 README       | 7 +++++++
 configure.ac | 2 +-
 2 files changed, 8 insertions(+), 1 deletion(-)

commit 94b335cee7281442f69e76cddea54b04ffe58c6e
Author: Ludovic Rousseau <ludovic.rousseau@free.fr>
Date:   Sat Jan 29 18:46:09 2022 +0100

    8 new ATRs

 smartcard_list.txt | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

commit a740e87d24d69855ca26639ba513228f9cb7e405
Author: Ludovic Rousseau <ludovic.rousseau@free.fr>
Date:   Fri Jan 28 10:43:17 2022 +0100

    pcsc_scan: handle Ctrl-C on macOS
    
    macOS does not support the plug & play mechanism "\\?PnP?\Notification"
    for SCardGetStatusChange() and uses instead an active polling loop over
    SCardListReaders().
    
    There is no SCardGetStatusChange() that can be cancled by SCardCancel()
    so we must handle the interruption by ourself.

 pcsc_scan.c | 6 ++++++
 1 file changed, 6 insertions(+)

commit ad38b2a420b1c808ae6e2a3fd9c9dcb3fbce5592
Author: Ludovic Rousseau <ludovic.rousseau@free.fr>
Date:   Fri Jan 21 14:53:43 2022 +0100

    10 new ATRs

 smartcard_list.txt | 39 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 39 insertions(+)

commit 2d685e74833d9d8ac01faef1d33b4a19155c4d62
Author: Philippe Teuwen <phil@teuwen.org>
Date:   Fri Jan 14 23:23:59 2022 +0100

    smartcard_list.txt: fix typos

 smartcard_list.txt | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

commit f0412761544dd6782b2c152a51ba760d0c946af6
Author: Ludovic Rousseau <ludovic.rousseau@free.fr>
Date:   Wed Jan 12 17:48:43 2022 +0100

    10 new ATRs

 smartcard_list.txt | 32 +++++++++++++++++++++++++++++++-
 1 file changed, 31 insertions(+), 1 deletion(-)

commit dd64def92639534aa5381306fcd3ffebfecfa53e
Author: Ludovic Rousseau <ludovic.rousseau@free.fr>
Date:   Tue Jan 4 19:09:58 2022 +0100

    10 new ATRs

 smartcard_list.txt | 36 ++++++++++++++++++++++++++++++++++++
 1 file changed, 36 insertions(+)

commit bffb77bafeaf99b8cd263145445632ba984dd7a7
Author: Ludovic Rousseau <ludovic.rousseau@free.fr>
Date:   Sat Dec 25 11:13:02 2021 +0100

    pcsc_scan: start the spinning thread before use
    
    The spinning mechanism can be used when waiting for the first reader.
    The thread must be started before that.

 pcsc_scan.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

commit c1d6a4b7a5dadcb4e329c87caf3944727a4ea53a
Author: Ludovic Rousseau <ludovic.rousseau@free.fr>
Date:   Fri Dec 24 20:33:52 2021 +0100

    pcsc_scan: fix spinning in quiet mode
    
    It was not possible to exit on Windows in quiet mode (-q)

 pcsc_scan.c | 2 +-

config.h
/* config.h.  Generated from config.h.in by configure.  */
/* config.h.in.  Generated from configure.ac by autoheader.  */

/* Define to 1 if you have the <inttypes.h> header file. */
#define HAVE_INTTYPES_H 1

/* Define to 1 if you have the <stdint.h> header file. */
#define HAVE_STDINT_H 1

/* Define to 1 if you have the <stdio.h> header file. */
#define HAVE_STDIO_H 1

/* Define to 1 if you have the <stdlib.h> header file. */
#define HAVE_STDLIB_H 1

/* Define to 1 if you have the <strings.h> header file. */
#define HAVE_STRINGS_H 1

/* Define to 1 if you have the <string.h> header file. */
#define HAVE_STRING_H 1

/* Define to 1 if you have the <sys/stat.h> header file. */
#define HAVE_SYS_STAT_H 1

/* Define to 1 if you have the <sys/time.h> header file. */
#define HAVE_SYS_TIME_H 1

/* Define to 1 if you have the <sys/types.h> header file. */
#define HAVE_SYS_TYPES_H 1

/* Define to 1 if you have the <time.h> header file. */
#define HAVE_TIME_H 1

/* Define to 1 if you have the <unistd.h> header file. */
#define HAVE_UNISTD_H 1

/* Define to 1 if `major', `minor', and `makedev' are declared in <mkdev.h>.
   */
/* #undef MAJOR_IN_MKDEV */

/* Define to 1 if `major', `minor', and `makedev' are declared in
   <sysmacros.h>. */
/* #undef MAJOR_IN_SYSMACROS */

/* Name of package */
#define PACKAGE "pcsc-tools"

/* Define to the address where bug reports for this package should be sent. */
#define PACKAGE_BUGREPORT ""

/* Define to the full name of this package. */
#define PACKAGE_NAME "pcsc-tools"

/* Define to the full name and version of this package. */
#define PACKAGE_STRING "pcsc-tools 1.6.0"

/* Define to the one symbol short name of this package. */
#define PACKAGE_TARNAME "pcsc-tools"

/* Define to the home page for this package. */
#define PACKAGE_URL ""

/* Define to the version of this package. */
#define PACKAGE_VERSION "1.6.0"

/* Define to 1 if all of the C90 standard headers exist (not just the ones
   required in a freestanding environment). This macro is provided for
   backward compatibility; new code need not use it. */
#define STDC_HEADERS 1

/* Version number of package */
#define VERSION "1.6.0"


include/libservo.h
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <wiringPi.h>

int run (void)
{
  if (wiringPiSetup() == -1)
    exit (1) ;

  pinMode(1,PWM_OUTPUT);
  pwmSetMode(PWM_MODE_MS);
  pwmSetClock(1920);
  pwmSetRange(200);
  
  // SI TOURNE DANS LE MAUVAIS SENS, METTRE (1,10)
  pwmWrite(1,20); //tourne dans un sens
  
  return 0;
}

int stop (void)
{
  if (wiringPiSetup() == -1)
    exit (1) ;

  pinMode(1,PWM_OUTPUT);
  pwmWrite(1,0); //stop
    
  return 0;
}

void delay(unsigned milliseconds)
{
    clock_t pause;
    clock_t start;
    
    pause = milliseconds * (CLOCKS_PER_SEC / 1000);
    start = clock();

    while ((clock() - start) < pause);
}


void rotationMoteur(float tempsSec)
{

    clock_t time1, time2;

    if(tempsSec == 0) return;

    //TODO : ajouter la fonction démarrage servo
    run();

    tempsSec = tempsSec *1000;

    delay(tempsSec);
    
    //TODO : ajouter la fonction arrêt servo
    stop();

}


include/libdistrib.h
#include "libdatabase.h"
#include "libservo.h"

int identiteChat(char tagAnimal[MAX_ATR_SIZE*3+1])
{
    // chat existe-t-il ?
    int idAnimal = getIdAnimal(tagAnimal);

    if(idAnimal == -1) idAnimal=newTag(tagAnimal);

    return idAnimal;
}

void delivreCroquettes(int idAnimal)
{
    struct objetDistrib objet = getQuantiteTotale(idAnimal);

    if(objet.qte == 0) return;

    printf("%s",objet.heure.total);

    float qteTotale = objet.qte;

    float qteParSeconde = getQuantiteParSec();

    printf("\nQuantité (g): %f\nDébit (g/sec): %f",qteTotale,qteParSeconde);

    float tempsRota = qteTotale/qteParSeconde;

    printf("\nTemps rota: %f\n", tempsRota);

    rotationMoteur(tempsRota);

    updateDB(idAnimal, objet);
}



void onTagOn(char tagAnimal[MAX_ATR_SIZE*3+1])
{

    int idAnimal = identiteChat(tagAnimal);
    delivreCroquettes(idAnimal);

}


pcsc_scan.c
/*
    Scan and print all the PC/SC devices available
    Copyright (C) 2001-2018  Ludovic Rousseau <ludovic.rousseau@free.fr>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
*/

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <string.h>
#include <signal.h>
typedef void (*sighandler_t)(int);
#include <sysexits.h>
#include <sys/time.h>
#include <pthread.h>
#include "include/libdistrib.h"

#ifdef __APPLE__
#include <PCSC/wintypes.h>
#include <PCSC/winscard.h>
#else
#include <winscard.h>
#endif

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#ifndef TRUE
#define TRUE 1
#define FALSE 0
#endif

typedef enum
{
	True = TRUE, False = FALSE
} Boolean;

#define TIMEOUT 3600*1000	/* 1 hour timeout */


/* command used to parse (on screen) the ATR */
#define ATR_PARSER "ATR_analysis"

#ifndef SCARD_E_NO_READERS_AVAILABLE
#define SCARD_E_NO_READERS_AVAILABLE 0x8010002E
#endif

#ifdef WIN32
const char *pcsc_stringify_error(DWORD rv)
{
	static char buffer[1024];

	if (! FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM,
		NULL,
		rv,
		MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), // default language
		buffer,
		sizeof buffer,
		NULL))
	{
		/* if FormatMessage() failed */
		snprintf(buffer, sizeof(buffer), "ERROR: 0x%08X\n", rv);
	}

	return buffer;
}
#define MAX_ATR_SIZE 33
#endif

#define print_pcsc_error(fct, rv) \
	fprintf(stderr, "%s%s: %s%s\n", red, fct, pcsc_stringify_error(rv), color_end)

#define test_rv(fct, rv, hContext) \
do { \
	if (rv != SCARD_S_SUCCESS) \
	{ \
		print_pcsc_error(fct, rv); \
		(void)SCardReleaseContext(hContext); \
		return -1; \
	} \
} while(0)

static void usage(const char *pname)
{
	printf("%s usage:\n\n\t%s [ -h | -V | -n | -r | -c | -s | -t secs ]\n\n", pname, pname);
	printf("\t\t  -h : this help\n");
	printf("\t\t  -V : print version number\n");
#ifndef WIN32
	printf("\t\t  -n : no ATR analysis\n");
#endif
	printf("\t\t  -r : only lists readers\n");
	printf("\t\t  -c : only lists cards\n");
	printf("\t\t  -s : stress mode\n");
	printf("\t\t  -q : quiet mode\n");
	printf("\t\t  -v : verbose mode (default)\n");
	printf("\t\t  -t secs : quit after secs seconds\n");
	printf("\n");
}

const char *blue = "";
const char *red = "";
const char *magenta = "";
const char *color_end = "";

const char *cub2 = "";
const char *cub3 = "";
const char *cpl = "";

time_t start_time;
_Atomic Boolean Interrupted = False;

typedef struct
{
	const char *pname;
	Boolean analyse_atr;
	Boolean stress_card;
	Boolean print_version;
	Boolean verbose;
	Boolean only_list_readers;
	Boolean only_list_cards;
	long maxtime; // in seconds
} options_t;

static options_t Options;

SCARDCONTEXT hContext;

static Boolean is_member(const char *  item, const char * list[])
{
	int i = 0;
	while (list[i] && 0 != strcmp(item, list[i]))
	{
		i++;
	}
	return list[i] != 0;
}

static void initialize_terminal(void)
{
	const char *color_terms[] = { "linux", "xterm", "xterm-color", "xterm-256color",
		"Eterm", "rxvt", "rxvt-unicode", "cygwin", 0};
	const char *no_ansi_cursor_terms[] = {"dumb", "emacs", 0};
	const char *term = getenv("TERM");

	/* do not use color if stdout is redirected */
	if (!isatty(fileno(stdout)) || (NULL == term))
	{
		term = "dumb";
	}
	if (is_member(term, color_terms))
	{
		blue = "\033[34m";
		red = "\033[31m";
		magenta = "\033[35m";
		color_end = "\033[0m";
	}
	if (is_member(term, no_ansi_cursor_terms))
	{
		cub2 = "\r"; /* use carriage return */
		cub3 = "\r";
		cpl = "\n"; /* can't do previous line,  let's go to the next line. */
	}
	else
	{
		cub2 = "\033[2D"; /* move back 2 characters */
		cub3 = "\033[3D"; /* move back 3 characters */
		cpl = "\033[";
	}
}
/* There should be no \033 beyond this line! */

static Boolean should_exit(void)
{
	if (Options.maxtime)
	{
		time_t t = time(NULL);
		if (t - start_time > Options.maxtime)
			return True;
	}

#ifdef WIN32
	if (GetKeyState(VK_SHIFT) & 0x80)
		return True;
#endif

	if (Interrupted)
		return True;

	return False;
}

typedef enum
{
	SpinDisabled = -2,
	SpinStopped = -1,
	SpinRunning = 0
} SpinState_t;

_Atomic SpinState_t spin_state = SpinStopped;

static void spin_start(void)
{
	spin_state = Options.verbose ? SpinRunning : SpinDisabled;
}

static void spin_stop(void)
{
	spin_state = SpinStopped;

	/* clean previous output */
	printf("%s %s ", cub2, cub2);
	fflush(stdout);
}

static void *spin_update(void *p)
{
	char patterns[] = {'-', '\\', '|', '/'};

	/* 100 ms wait */
	struct timespec wait_time = {.tv_sec = 0, .tv_nsec = 1000*1000*100};

	(void)p;

again:
	/* wait untill spinning starts */
	do
	{
		if (should_exit())
		{
			if (SpinDisabled == spin_state)
			{
				SCardCancel(hContext);
				pthread_exit(NULL);
			}
		}

		nanosleep(&wait_time, NULL);
	} while (spin_state < 0);

	do
	{
		char c = patterns[spin_state];

		if (should_exit())
		{
			SCardCancel(hContext);
			pthread_exit(NULL);
		}

		spin_state++;
		if (spin_state >= (int)sizeof patterns)
			spin_state = SpinRunning;
		printf("%s %c ", cub3, c);
		fflush(stdout);

		nanosleep(&wait_time, NULL);
	} while (spin_state >= SpinRunning);

	goto again;

	return NULL;
}

static void user_interrupt_signal_handler(int signal)
{
	(void)signal;
	Interrupted = True;
}

static void initialize_signal_handlers(void)
{
	signal(SIGINT, user_interrupt_signal_handler);
}


static void initialize_options(options_t *options, const char *pname)
{
	options->pname = pname;
#ifdef WIN32
	options->analyse_atr = False;
#else
	options->analyse_atr = True;
#endif
	options->stress_card = False;
	options->print_version = False;
	options->maxtime = 0;
	options->verbose = True;
	options->only_list_readers = False;
	options->only_list_cards = False;
}

#define OPTIONS_BASE "Vhvrcst:q"
#ifdef WIN32
#define OPTIONS OPTIONS_BASE
#else
#define OPTIONS OPTIONS_BASE "n"
#endif

static int parse_options(int argc, char *argv[], options_t *options)
{
	const char *pname = argv[0];
	int opt;
	initialize_options(options, pname);
	while ((opt = getopt(argc, argv, OPTIONS)) != EOF)
	{
		switch (opt)
		{
			case 'n':
				options->analyse_atr = False;
				break;

			case 'V':
				options->print_version = True;
				break;

			case 'v':
				options->verbose = True;
				break;

			case 'r':
				options->only_list_readers = True;
				options->verbose = False;
				break;

			case 'c':
				options->only_list_cards = True;
				options->verbose = False;
				options->analyse_atr = False;
				break;

			case 's':
				options->stress_card = True;
				break;

			case 't':
				options->maxtime = atol(optarg);
				break;

			case 'q':
				options->verbose = False;
				break;

			case 'h':
				usage(pname);
				exit(EX_OK);

			default:
				usage(pname);
				exit(EX_USAGE);
		}
	}
	if (argc - optind != 0)
	{
		fprintf(stderr, "%s error: superfluous arguments: %s\n", pname, argv[optind]);
		usage(pname);
		exit(EX_USAGE);
	}
	return EX_OK;
}

static LONG stress(SCARDCONTEXT hContext2, const char *readerName)
{
	LONG rv, ret_rv = SCARD_S_SUCCESS;
	SCARDHANDLE hCard;
	DWORD dwActiveProtocol;
	const SCARD_IO_REQUEST *pioSendPci;

	if (Options.verbose)
	{
		printf("Stress card in reader: %s\n\n", readerName);
	}
	else
		printf("\n");
	rv = SCardConnect(hContext2, readerName, SCARD_SHARE_SHARED,
			SCARD_PROTOCOL_T0 | SCARD_PROTOCOL_T1, &hCard, &dwActiveProtocol);
	if (rv != SCARD_S_SUCCESS)
	{
		print_pcsc_error("SCardConnect", rv);
		return rv;
	}

	/* Select Master File */
	BYTE pbSendBuffer[] = {0, 0xA4, 0, 0, 2, 0x3F, 0};
	BYTE pbRecvBuffer[256+2];
	DWORD dwSendLength, dwRecvLength;
	struct timeval time_start, time_end;

	switch (dwActiveProtocol)
	{
		case SCARD_PROTOCOL_T0:
			pioSendPci = SCARD_PCI_T0;
			break;
		case SCARD_PROTOCOL_T1:
			pioSendPci = SCARD_PCI_T1;
			break;
		case SCARD_PROTOCOL_RAW:
			pioSendPci = SCARD_PCI_RAW;
			break;
		default:
			fprintf(stderr, "Unknown protocol\n");
			return -1;
	}

	gettimeofday(&time_start, NULL);


#define COUNT 100
	size_t count;
	for (count=0; count<COUNT; count++)
	{
		printf("%sFAPDU n°: %ld\n", cpl, count);
		dwSendLength = sizeof(pbSendBuffer);
		dwRecvLength = sizeof(pbRecvBuffer);
		rv = SCardTransmit(hCard, pioSendPci, pbSendBuffer, dwSendLength,
			NULL, pbRecvBuffer, &dwRecvLength);
		if (rv != SCARD_S_SUCCESS)
		{
			print_pcsc_error("SCardTransmit", rv);
			ret_rv = rv;
			break;
		}

		if (Interrupted)
			exit(EX_OSERR);
	}

	gettimeofday(&time_end, NULL);
	struct timeval r;
	r.tv_sec = time_end.tv_sec - time_start.tv_sec;
	r.tv_usec = time_end.tv_usec - time_start.tv_usec;
	if (r.tv_usec < 0)
	{
		r.tv_sec--;
		r.tv_usec += 1000000;
	}
	long delta = r.tv_sec * 1000000 + r.tv_usec;
	printf("Total time: %ld µs\n", delta);
	if (0 == count)
		count = 1;
	printf("%f APDU/s\n", 1000000. / (delta / count));

	rv = SCardDisconnect(hCard, SCARD_UNPOWER_CARD);
	if (rv != SCARD_S_SUCCESS)
	{
		print_pcsc_error("SCardDisconnect", rv);
		ret_rv = rv;
	}

	return ret_rv;
}

static void print_readers(const char **readers, int nbReaders)
{
	int i = 0;
	for (i = 0;i < nbReaders; i++)
	{
		printf("%s%d: %s%s\n", blue, i, readers[i], color_end);
	}
}

static void print_version(void)
{
	printf("PC/SC device scanner\n");
	printf("V %s (c) 2001-2018, Ludovic Rousseau <ludovic.rousseau@free.fr>\n",
			PACKAGE_VERSION);
}

int main(int argc, char *argv[])
{
	int current_reader;
#ifdef __APPLE__
	uint32_t rv;
#else
	LONG rv;
#endif
	SCARD_READERSTATE *rgReaderStates_t = NULL;
	SCARD_READERSTATE rgReaderStates[1];
	DWORD dwReaders = 0, dwReadersOld;
	LPSTR mszReaders = NULL;
	char *ptr = NULL;
	const char **readers = NULL;
	int nbReaders, i;
	char atr[MAX_ATR_SIZE*3+1];	/* ATR in ASCII */
	char atr_command[sizeof(atr)+sizeof(ATR_PARSER)+2+1];
	int pnp = TRUE;
	pthread_t spin_pthread;

	// Variables chronomètre
	struct timespec tstart={0,0}, tend={0,0};
	Boolean firstFlag=False;

	char tagChat[MAX_ATR_SIZE*3+1];

	start_time = time(NULL);
	initialize_terminal();
	if (0 != parse_options(argc, argv, &Options))
	{
		exit(EX_USAGE);
	}
	if (Options.print_version)
	{
		print_version();
		exit(EX_OK);
	}

#ifdef WIN32
	if (Options.verbose)
	{
		printf("%sPress shift key to quit%s\n", magenta, color_end);
	}
#else
	initialize_signal_handlers();
#endif

	rv = SCardEstablishContext(SCARD_SCOPE_SYSTEM, NULL, NULL, &hContext);
	test_rv("SCardEstablishContext", rv, hContext);

	rgReaderStates[0].szReader = "\\\\?PnP?\\Notification";
	rgReaderStates[0].dwCurrentState = SCARD_STATE_UNAWARE;

	rv = SCardGetStatusChange(hContext, 0, rgReaderStates, 1);
	if (rgReaderStates[0].dwEventState & SCARD_STATE_UNKNOWN)
	{
		if (Options.verbose)
		{
			printf("%sPlug'n play reader name not supported. Using polling every %d ms.%s\n", magenta, TIMEOUT, color_end);
		}
		pnp = FALSE;
	}
	else
	{
		if (Options.verbose)
		{
			printf("%sUsing reader plug'n play mechanism%s\n", magenta, color_end);
		}
	}

	/* start spining thread */
	rv = pthread_create(&spin_pthread, NULL, spin_update, NULL);

get_readers:
	/* free memory possibly allocated in a previous loop */
	if (NULL != readers)
	{
		free(readers);
		readers = NULL;
	}

	if (NULL != rgReaderStates_t)
	{
		free(rgReaderStates_t);
		rgReaderStates_t = NULL;
	}

	/* Retrieve the available readers list.
	 *
	 * 1. Call with a null buffer to get the number of bytes to allocate
	 * 2. malloc the necessary storage
	 * 3. call with the real allocated buffer
	 */
	if (Options.verbose)
	{
		printf("%sScanning present readers...%s\n", red, color_end);
	}
	rv = SCardListReaders(hContext, NULL, NULL, &dwReaders);
	if (rv != SCARD_E_NO_READERS_AVAILABLE)
		test_rv("SCardListReaders", rv, hContext);

	dwReadersOld = dwReaders;

	/* if non NULL we came back so free first */
	if (mszReaders)
	{
		free(mszReaders);
		mszReaders = NULL;
	}

	mszReaders = malloc(sizeof(char)*dwReaders);
	if (mszReaders == NULL)
	{
		fprintf(stderr, "%s: malloc: not enough memory\n", Options.pname);
		exit(EX_OSERR);
	}

	*mszReaders = '\0';
	rv = SCardListReaders(hContext, NULL, mszReaders, &dwReaders);

	/* Extract readers from the null separated string and get the total
	 * number of readers */
	nbReaders = 0;
	ptr = mszReaders;
	while (*ptr != '\0')
	{
		ptr += strlen(ptr)+1;
		nbReaders++;
	}

	if (SCARD_E_NO_READERS_AVAILABLE == rv || 0 == nbReaders)
	{
		if (Options.only_list_cards || Options.only_list_readers)
		{
			printf("No reader found.\n");
			return EX_OK;
		}

		if (Options.verbose)
		{
			printf("%sWaiting for the first reader...%s   ", red, color_end);
			fflush(stdout);
		}

		if (pnp)
		{
			rgReaderStates[0].szReader = "\\\\?PnP?\\Notification";
			rgReaderStates[0].dwCurrentState = SCARD_STATE_UNAWARE;

			spin_start();
			do
			{
				rv = SCardGetStatusChange(hContext, TIMEOUT, rgReaderStates, 1);
			}
			while (SCARD_E_TIMEOUT == rv);
			spin_stop();

			test_rv("SCardGetStatusChange", rv, hContext);
		}
		else
		{
			rv = SCARD_S_SUCCESS;
			spin_start();
			while ((SCARD_S_SUCCESS == rv) && (dwReaders == dwReadersOld))
			{
				rv = SCardListReaders(hContext, NULL, NULL, &dwReaders);
				if (SCARD_E_NO_READERS_AVAILABLE == rv)
					rv = SCARD_S_SUCCESS;
				sleep(1);

				if (should_exit())
				{
					printf("\n");
					exit(EX_OK);
				}
			}
			spin_stop();
		}
		if (Options.verbose)
		{
			printf("found one\n");
		}
		goto get_readers;
	}
	else
		test_rv("SCardListReader", rv, hContext);

	/* allocate the readers table */
	readers = calloc(nbReaders+1, sizeof(char *));
	if (NULL == readers)
	{
		fprintf(stderr, "%s: Not enough memory for readers table\n", Options.pname);
		exit(EX_OSERR);
	}

	/* fill the readers table */
	nbReaders = 0;
	ptr = mszReaders;
	while (*ptr != '\0')
	{
		readers[nbReaders] = ptr;
		ptr += strlen(ptr)+1;
		nbReaders++;
	}


	// MONTRE LE READER UTILISE

	if (! Options.only_list_cards)
		print_readers(readers, nbReaders);
	if (Options.only_list_readers)
	{
		exit(EX_OK);
	}

	/* allocate the ReaderStates table */
	rgReaderStates_t = calloc(nbReaders+1, sizeof(* rgReaderStates_t));
	if (NULL == rgReaderStates_t)
	{
		fprintf(stderr, "%s: Not enough memory for readers states\n", Options.pname);
		exit(EX_OSERR);
	}

	/* Set the initial states to something we do not know
	 * The loop below will include this state to the dwCurrentState
	 */
	for (i=0; i<nbReaders; i++)
	{
		rgReaderStates_t[i].szReader = readers[i];
		rgReaderStates_t[i].dwCurrentState = SCARD_STATE_UNAWARE;
		rgReaderStates_t[i].cbAtr = sizeof rgReaderStates_t[i].rgbAtr;
	}

	/* If Plug and Play is supported by the PC/SC layer */
	if (pnp)
	{
		rgReaderStates_t[nbReaders].szReader = "\\\\?PnP?\\Notification";
		rgReaderStates_t[nbReaders].dwCurrentState = SCARD_STATE_UNAWARE;
		nbReaders++;
	}

#ifdef WIN32
	int oldNbReaders;
	int oldNbReaders_init = FALSE;
#endif


	 // ON ECOUTE


	spin_start();

	/* Wait endlessly for all events in the list of readers
	 * We only stop in case of an error
	 */
	rv = SCardGetStatusChange(hContext, TIMEOUT, rgReaderStates_t, nbReaders);

	spin_stop();





	// BOUCLE A MODIFIER


	while ((rv == SCARD_S_SUCCESS) || (rv == SCARD_E_TIMEOUT))
	{
		time_t t;

		if (pnp)
		{
			/* check if the number of readers has changed */
#ifdef WIN32
			LONG newNbReaders = rgReaderStates_t[nbReaders-1].dwEventState >> 16;
			if (! oldNbReaders_init)
			{
				oldNbReaders = newNbReaders;
				oldNbReaders_init = TRUE;
			}
			if (newNbReaders != oldNbReaders)
#else
			if (rgReaderStates_t[nbReaders-1].dwEventState &
					SCARD_STATE_CHANGED)
#endif
			{
				goto get_readers;
			}
		}
		else
		{
			/* A new reader appeared? */
			if ((SCardListReaders(hContext, NULL, NULL, &dwReaders)
				== SCARD_S_SUCCESS) && (dwReaders != dwReadersOld))
			{
				goto get_readers;
			}
		}

		if (rv != SCARD_E_TIMEOUT)
		{
			/* Timestamp the event as we get notified */
			t = time(NULL);
			printf("\n%s", ctime(&t));
		}

		/* Now we have an event, check all the readers in the list to see what
		 * happened */
		for (current_reader=0; current_reader < nbReaders; current_reader++)
		{
#if defined(__APPLE__) || defined(WIN32)
			if (rgReaderStates_t[current_reader].dwCurrentState ==
				rgReaderStates_t[current_reader].dwEventState)
				continue;
#endif

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_CHANGED)
			{
				/* If something has changed the new state is now the current
				 * state */
				rgReaderStates_t[current_reader].dwCurrentState =
					rgReaderStates_t[current_reader].dwEventState;
			}
			else
				/* If nothing changed then skip to the next reader */
				continue;

			/* From here we know that the state for the current reader has
			 * changed because we did not pass through the continue statement
			 * above.
			 */

			/* Dump the full current state */
			printf("  Card state: %s", red);

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_IGNORE)
				printf("Ignore this reader, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_UNKNOWN)
			{
				printf("Unknown\n");
				goto get_readers;
			}

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_UNAVAILABLE)
				printf("Status unavailable, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_EMPTY){

					printf("Card removed, ");
					
				}
				
					

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_PRESENT){

					printf("Card inserted, ");

				}
				

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_ATRMATCH)
				printf("ATR matches card, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_EXCLUSIVE)
				printf("Exclusive Mode, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_INUSE)
				printf("Shared Mode, ");

			if (rgReaderStates_t[current_reader].dwEventState &
				SCARD_STATE_MUTE)
				printf("Unresponsive card, ");

			printf("%s\n", color_end);

			/* force display */
			fflush(stdout);

			/* Also dump the ATR if available */
			if (rgReaderStates_t[current_reader].cbAtr > 0)
			{
				printf("  ATR: ");

				if (rgReaderStates_t[current_reader].cbAtr)
				{
					unsigned int j;

					for (j=0; j<rgReaderStates_t[current_reader].cbAtr; j++)
						sprintf(&atr[j*3], "%02X ",
							rgReaderStates_t[current_reader].rgbAtr[j]);

					atr[j*3-1] = '\0';
				}
				else
					atr[0] = '\0';


				strcpy(tagChat,atr);
				printf("%s%s%s\n", magenta, atr, color_end);

				// Programme distribution et MAJ BDD
				onTagOn(tagChat);

				/* force display */
				fflush(stdout);

				if (Options.analyse_atr)
				{
					printf("\n");

					sprintf(atr_command, ATR_PARSER " '%s'", atr);
					if (system(atr_command))
						perror(atr_command);
				}
			}

			LONG state = rgReaderStates_t[current_reader].dwEventState;
			if (state & SCARD_STATE_PRESENT
				&& !(state & SCARD_STATE_MUTE)
				&& Options.stress_card)
			{
				do
				{
					rv = stress(hContext, rgReaderStates_t[current_reader].szReader);
				} while (SCARD_S_SUCCESS == rv);

				rgReaderStates_t[current_reader].dwCurrentState = SCARD_STATE_UNAWARE;
			}
		} /* for */

		if (Options.only_list_cards)
			break;

		spin_start();

		rv = SCardGetStatusChange(hContext, TIMEOUT, rgReaderStates_t,
			nbReaders);

		spin_stop();
	} /* while */

	/* A reader disappeared */
	if (SCARD_E_UNKNOWN_READER == rv)
		goto get_readers;

	/* If we get out the loop, GetStatusChange() was unsuccessful */
	if (rv != SCARD_E_CANCELLED)
		test_rv("SCardGetStatusChange", rv, hContext);
	else
	{
		printf("%s", cub3);
		fflush(stdout);
		pthread_join(spin_pthread, NULL);
	}

	/* We try to leave things as clean as possible */
	rv = SCardReleaseContext(hContext);
	test_rv("SCardReleaseContext", rv, hContext);

	/* free memory possibly allocated */
	if (NULL != readers)
		free(readers);
	if (NULL != rgReaderStates_t)
		free(rgReaderStates_t);

	return EX_OK;
}
